### PR TITLE
exchange refinement flags

### PR DESCRIPTION
**Pull request by quangx**
Communicate the refinement/coarsen flags in p::d::Triangulation::prepare_coarsening_and_refinement to get consistent refinement independent of the number of processors involved in the computation.

We have been doing this as a work-around in ASPECT for a long time:
https://github.com/geodynamics/aspect/blob/5bf882496a69216385a3ca28715752a1b231cc20/source/simulator/helper_functions.cc#L546-L550
This PR takes the exchange_refinement_flags function from ASPECT and calls it in prepare_coarsening_and_refinement.

Additionally:
1. Update test output where the number of cells change slightly.
2. Use the correct sequential prepare_coarsening_and_refinement function in copy_local_forest_to_triangulation
3. Only do the artificial 2:1 periodic enforcement in copy_local_forest_to_triangulation, which is percisely where we need it.

### CODE REVIEW

File: source/distributed/tria.cc [hunk 1]
```diff
@@ -3228,6 +3245,39 @@ namespace parallel
     }
 
 
+    template <int dim, int spacedim>
+    void
+    exchange_refinement_flags(Triangulation<dim, spacedim> &tria)
```
Comment (bangerth) [inline source/distributed/tria.cc:3250]: I don't see the declaration of this function.

File: source/distributed/tria.cc [hunk 2]
```diff
@@ -3228,6 +3245,39 @@ namespace parallel
...
+        cell->clear_coarsen_flag();
+        cell->clear_refine_flag();
+        if (flag == 1)
+          cell->set_refine_flag();
+        else if (flag == 2)
+          cell->set_coarsen_flag();
+      };
+
+      GridTools::exchange_cell_data_to_ghosts<std::uint8_t,
+                                              Triangulation<dim, spacedim>>(
```
Comment (bangerth) [inline source/distributed/tria.cc:3278]: I suspect that you can write this as
```suggestion
      GridTools::exchange_cell_data_to_ghosts<std::uint8_t>(
```

**Review by bangerth — COMMENTED** [review]
This is clearly an incompatible change. Can you explain what problem this patch is solving, and why it is necessary?

**Comment by bangerth** [comment]
My understanding was always that refinement flags on ghost/artificial cells are simply ignored (or that in fact we test that they are not set). If the refinement algorithm internally needs them, then it internally needs to communicate them. Is this not true?

**Comment by tjhei** [comment]
> My understanding was always that refinement flags on ghost/artificial cells are simply ignored

This is not the case. As I said, this is what we do in ASPECT for a good reason:
https://github.com/geodynamics/aspect/blob/5bf882496a69216385a3ca28715752a1b231cc20/source/simulator/helper_functions.cc#L546-L550

> If the refinement algorithm internally needs them, then it internally needs to communicate them

This is exactly what this patch is proposing! The flag smoothing is done in the sequential prepare_coarsening_and_refinement function and necessarily depends on flags on ghost cells. There is no communication happening there, but this patch proposes fixing exactly that by introducing it in the parallel version, which internally calls the sequential one. Note that this has been broken/inconsistent for many, many years...

Note: This patch is in a sequence to address several other problems regarding refinement, mesh smoothing, and periodic boundary conditions like #1655 #7053 #10604 #16141

File: source/distributed/tria.cc (hunk 1 above)
Comment (tjhei) [reply source/distributed/tria.cc:3250]: @quangx Can you move this into the anonymous namespace higher up in the file, maybe here: https://github.com/dealii/dealii/blob/29f9da0a3412c3530b2f1b93c0f59ab6d909546f/source/distributed/tria.cc#L568 ?

**Comment by tjhei** [comment]
> Would it be possible to add a check that upon calling `execute_c_and_r()`, nothing is set on ghost cells?

We could do that but this would break a lot of existing tests and likely other code, so I am not sure this is a good idea.

**Comment by bangerth** [comment]
I'm ok leaving this for a separate PR. I think it's the right thing to do, because setting things on ghost cells suggests that that is meaningful when in reality it is not.

You'll have to adjust a few more tests to make the CI happy, though.

**Comment by bangerth** [comment]
There is one test that fails: `matrix_free/cell_categorization.mpirun=7.debug`. Can you take a look?

**Review by bangerth — COMMENTED** [review]
To reiterate, this patch makes sense to me once the one test succeeds. 

I have one question, though: In the existing implementation, `prepare_for_c_and_r()` has a loop that iterates out flags until we have reached convergence. This patch moves this loop to the `copy_local_forest_to_triangulation()` function. But this is only a change that is correct if we know that the iteration would terminate after the first iteration for all calls of `prepare_for_c_and_r()` that originate *outside* of `copy_local_forest_to_triangulation()`. Do we know this to be true?

I'll note that this last point has nothing really to do with the exchange of refinement flags to ghosts -- that part of the patch is clearly correct.

**Comment by marcfehling** [comment]
Related to #8647.

**Comment by tjhei** [comment]
> To reiterate, this patch makes sense to me once the one test succeeds.

Great. I updated the test results for different vectorization levels, hopefully that works now.

> In the existing implementation, prepare_for_c_and_r() has a loop that iterates out flags until we have reached convergence

Yes, that is correct. I don't have 100% confidence that something else might go wrong, but the only reason for this loop in the old code is to make sure that there are no 2+ level of refinement difference to artificial cells (which are consequently ignored anyways).

Finally, we have a followup patch coming soon that adds a loop around the "exchange" call to improve support for mesh smoothing and other problems in parallel.

**Comment by bangerth** [comment]
Sounds good. Let me know once you've got the test running!

**Comment by tjhei** [comment]
Tests are finally passing.

**Review by bangerth — APPROVED** [review]

**Comment by tjhei** [comment]
Thank you @quangx !

@vyushut What is the status of the next step?

**Comment by marcfehling** [comment]
I believe #8647 can now be closed as well?

And the workaround in #8637 has become obsolete? Maybe we should turn the workaround into an assertion now.

**Pull request merged**